### PR TITLE
⚙️ Add providers and processors

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: src/Component/State/CallableProcessor.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    public function __construct(private ContainerInterface $locator)
+    {
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function process(mixed $data, Operation $operation, Context $context)
+    {
+        if (!($processor = $operation->getProcessor())) {
```
Comment (lchrusciel) [inline src/Component/State/CallableProcessor.php]:
```suggestion
        $processor = $operation->getProcessor()
        
        if (is_null($processor)) {
```

File: src/Component/State/CallableProcessor.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        if (\is_callable($processor)) {
+            return $processor($data, $operation, $context);
+        }
+
+        if (!$this->locator->has($processor)) {
+            throw new \RuntimeException(sprintf('Processor "%s" not found on operation "%s"', $processor, $operation->getName() ?? ''));
+        }
+
+        /** @var ProcessorInterface $processorInstance */
```
Comment (lchrusciel) [inline src/Component/State/CallableProcessor.php]: Perhaps we can assert on that?

File: src/Component/State/CallableProvider.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+final class CallableProvider implements ProviderInterface
+{
+    public function __construct(private ContainerInterface $locator)
+    {
+    }
+
+    public function provide(Operation $operation, Context $context): object|iterable|null
+    {
+        if (!($provider = $operation->getProvider())) {
```
Comment (lchrusciel) [inline src/Component/State/CallableProvider.php]:
```suggestion
        $provider = $operation->getProvider()
        
        if (is_null($provider)) {        
        
```

File: src/Component/State/CallableProvider.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+        if (\is_callable($provider)) {
+            return $provider($operation, $context);
+        }
+
+        if (!$this->locator->has($provider)) {
+            throw new \RuntimeException(sprintf('Provider "%s" not found on operation "%s"', $provider, $operation->getName() ?? ''));
+        }
+
+        /** @var ProviderInterface $providerInstance */
```
Comment (lchrusciel) [inline src/Component/State/CallableProvider.php]: Perhaps we can assert on that?

File: src/Component/State/ProcessorInterface.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * @experimental
+ */
+interface ProcessorInterface
+{
+    /**
+     * Handle the state.
+     *
+     * @return mixed
+     */
+    public function process(mixed $data, Operation $operation, Context $context);
```
Comment (lchrusciel) [inline src/Component/State/ProcessorInterface.php]:
```suggestion
     */
    public function process(mixed $data, Operation $operation, Context $context): mixed;
```

File: src/Component/State/ProviderInterface.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace Sylius\Component\Resource\State;
+
+use Sylius\Component\Resource\Context\Context;
+use Sylius\Component\Resource\Metadata\Operation;
+
+/**
+ * Retrieves data from a persistence layer.
+ *
+ * @experimental
```
Comment (lchrusciel) [inline src/Component/State/ProviderInterface.php:22]: Is it the only class in this PR that should be experimental? Or should we remove it/add it everywhere?

File: src/Component/spec/State/CallableProcessorSpec.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    function it_calls_process_method_from_operation_processor_as_callable(): void
+    {
+        $operation = new Create(processor: [ProcessorWithCallable::class, 'process']);
+        $context = new Context();
+
+        $this->process([], $operation, $context)->shouldHaveType(\stdClass::class);
+    }
+
+    function it_returns_null_if_operation_has_no_provider(): void
```
Comment (lchrusciel) [inline src/Component/spec/State/CallableProcessorSpec.php]:
```suggestion
    function it_returns_null_if_operation_has_no_processor(): void
```

File: src/Component/State/ProviderInterface.php (hunk 6 above)
Comment (loic425) [reply src/Component/State/ProviderInterface.php:22]: I add it on a lot of classes (on Operations for example), but indeed, maybe that'd be better to add this on every new classes on this PR.

Comment (loic425) [reply src/Component/State/ProviderInterface.php:22]: I added experimental tags on CallableProvider and CallableProcessor from this PR.

File: src/Component/State/CallableProvider.php (hunk 4 above)
Comment (loic425) [reply src/Component/State/CallableProvider.php]: done

File: src/Component/State/CallableProcessor.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+use Psr\Container\ContainerInterface;
+use Sylius\Component\Resource\Context\Context;
+use Sylius\Component\Resource\Metadata\Operation;
+use Webmozart\Assert\Assert;
+
+/**
+ * @experimental
+ */
+final class CallableProcessor implements ProcessorInterface
```
Comment (lchrusciel) [inline src/Component/State/CallableProcessor.php:24]: This class is not responsible for callable processors only
```suggestion
final class Processor implements ProcessorInterface
```
or 
```suggestion
final class ProcessorExecutor implements ProcessorInterface
```

File: src/Component/State/CallableProvider.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+use Psr\Container\ContainerInterface;
+use Sylius\Component\Resource\Context\Context;
+use Sylius\Component\Resource\Metadata\Operation;
+use Webmozart\Assert\Assert;
+
+/**
+ * @experimental
+ */
+final class CallableProvider implements ProviderInterface
```
Comment (lchrusciel) [inline src/Component/State/CallableProvider.php:24]: as above
```suggestion
final class Provider implements ProviderInterface
```

File: src/Component/State/CallableProcessor.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        if (\is_callable($processor)) {
+            return $processor($data, $operation, $context);
+        }
+
+        if (!$this->locator->has($processor)) {
+            throw new \RuntimeException(sprintf('Processor "%s" not found on operation "%s"', $processor, $operation->getName() ?? ''));
+        }
+
+        $processorInstance = $this->locator->get($processor);
+        Assert::isInstanceOf($processorInstance, ProcessorInterface::class);
```
Comment (lchrusciel) [inline src/Component/State/CallableProcessor.php:50]: These two exceptions are not tested

File: src/Component/State/CallableProvider.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        if (\is_callable($provider)) {
+            return $provider($operation, $context);
+        }
+
+        if (!$this->locator->has($provider)) {
+            throw new \RuntimeException(sprintf('Provider "%s" not found on operation "%s"', $provider, $operation->getName() ?? ''));
+        }
+
+        $providerInstance = $this->locator->get($provider);
+        Assert::isInstanceOf($providerInstance, ProviderInterface::class);
```
Comment (lchrusciel) [inline src/Component/State/CallableProvider.php:47]: These two exceptions are not tested

**Review by lchrusciel — APPROVED** [review]

**Pull request merged**